DP4-2 [Admin] Create/Update product

koleshkos opened the pull request:
- Added button "Create"/"Edit"
- Validates for Product model:
  - Name
  - Image
  - Code
  - Price
  - Description
- Added Bootstrap

DmitryGolubko commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +12,36 @@
+      <nav class="navbar navbar-expand-lg navbar-dark bg-dark">
+        <div class="container">
+          <ul class="navbar-nav mr-auto">
+            <a class="navbar-brand" href="/">MℽSℍOℙ</a>
```
what a strange symbols in `MℽSℍOℙ`, use normal ones

DmitryGolubko commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +12,36 @@
+        <div class="container">
+          <ul class="navbar-nav mr-auto">
+            <a class="navbar-brand" href="/">MℽSℍOℙ</a>
+            <li><%= link_to 'Product', products_path, class: 'nav-link px-2 text-white'%></li>
```
Is it a link to a product list? then we should name it 'Products'

DmitryGolubko commented on app/views/layouts/application.html.erb:
```diff
@@ -11,6 +12,36 @@
+          <div class="footer-left mt-2 mb-2">
+            <%= link_to 'FAQ', root_path, class: 'text-decoration-none text-white-50'%>
+            <%= link_to 'Help', root_path, class: 'text-decoration-none text-white-50'%>
+            <%= link_to 'About', root_path, class: 'text-decoration-none text-white-50'%>
```
Are we gonna have these pages? remove if not

DmitryGolubko commented on app/views/products/new.html.haml:
```diff
@@ -0,0 +1,20 @@
+    = f.label :description
+    = f.text_field :description, class: 'form-control'
+    %br
+    = f.submit 'Save', class: 'btn btn-primary'
```
What's the difference between `new` and `edit` partials? Let's create separate partial 'views/products/_form' for form only and render it for new and edit actions.

DmitryGolubko commented on app/views/products/show.html.haml:
```diff
@@ -0,0 +1,14 @@
+%br
+= @product.code
+%br
+= @product.price
```
Let's use `number_to_currency` helper there for better price view
https://apidock.com/rails/ActionView/Helpers/NumberHelper/number_to_currency

DmitryGolubko commented on db/migrate/20220131094309_create_products.rb:
```diff
@@ -0,0 +1,15 @@
+      t.string :name
+      t.string :image
+      t.string :code
+      t.integer :price
```
what if price will be with cents? let's use decimal here with `scale: 2`
https://nelsonfigueroa.dev/handling-decimal-precision-in-rails/

DmitryGolubko commented on app/controllers/products_controller.rb:
```diff
@@ -1,5 +1,54 @@
+  end
+
+  def show
+    @product = Product.find(params[:id])
```
let's move this finder to private method `get_product` which will be called for show, edit, update and destroy methods, use `before_filter` for this

DmitryGolubko reviewed (changes requested):
Changes requested.

TODO: 
1. Remove all temp images from the github, github is not intended for storing images
2. Migrate image storing to AWS
3. Write specs for your model and controller, check that everything is working correct

koleshkos commented:
- Removed all temp images from the github
- Migrated image storing to AWS
- Added before_action :set_product
- Used  decimal  with scale: 2 for price
- Used number_to_currency helper  for better price view
- Added partials form for new and edit
- Changed link name to 'Products'
- Removed unnecessary links in footer
- Fixed name project in header

Next step:
 - Writing specs for  model and controller

DmitryGolubko commented on spec/controllers/products_controller_spec.rb:
```diff
@@ -0,0 +1,104 @@
+require 'rails_helper'
+
+RSpec.describe ProductsController, type: :controller do
+  let(:product) { Product.create(name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product') }
```
move product creation to factories, use factory_bot and faker gems.
Do the same for products with bad data using traits

DmitryGolubko commented on spec/controllers/products_controller_spec.rb:
```diff
@@ -0,0 +1,104 @@
+
+    it 'with bad data' do
+      post :create, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: 1000, price: '12345RE', description: 'My Product' } }
+      expect(response).not_to be_redirect
```
you also have to check that the new product is not created

DmitryGolubko commented on spec/controllers/products_controller_spec.rb:
```diff
@@ -0,0 +1,104 @@
+
+    it 'returns a success response' do
+      post :create, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product' } }
+      expect(response).to be_redirect
```
you also have to check that the new product is created

DmitryGolubko commented on spec/controllers/products_controller_spec.rb:
```diff
@@ -0,0 +1,104 @@
+  end
+
+  describe 'PATCH update' do
+    it 'updates the product and redirects' do
+      patch :update, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product' } }
+      expect(response).to be_redirect
+    end
+
+    it 'with bad data' do
+      patch :update, params: { id: product, product: { name: 'Product', image: 'product.jpg', code: 1000, price: '12345RE', description: 'My Product' } }
+      expect(response).not_to be_redirect
+    end
```
the same as for creation

DmitryGolubko commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,78 @@
+require 'rails_helper'
+
+RSpec.describe Product, type: :model do
+  subject { Product.new(name: 'Product', image: 'product.jpg', code: '12345RE', price: 1000, description: 'My Product') }
```
the same move product creating to factories

DmitryGolubko reviewed (changes requested):
couple of changes

DmitryGolubko commented on spec/controllers/products_controller_spec.rb:
```diff
@@ -0,0 +1,134 @@
+    product = FactoryBot.create(:product)
+
+    it 'returns a success response' do
+      get :new, params: { id: product }
```
Do we really need id in new template?

DmitryGolubko reviewed (changes requested):
check usage of id in new template and one more comment left about creating products
